agent/consul/state/catalog.go

```diff
@@ -55,10 +56,107 @@ func nodesTableSchema() *memdb.TableSchema {
 	}
 }
 
+// terminatingGatewayServicesTableSchema returns a new table schema used to store information
+// about services associated with terminating gateways.
+func terminatingGatewayServicesTableSchema() *memdb.TableSchema {
+	return &memdb.TableSchema{
+		Name: "terminating-gateway-services",
+		Indexes: map[string]*memdb.IndexSchema{
+			"id": {
+				Name:         "id",
+				AllowMissing: false,
+				Unique:       true,
+				Indexer: &memdb.CompoundIndex{
+					Indexes: []memdb.Indexer{
+						&ServiceIDIndex{
+							Field: "Gateway",
+						},
+						&ServiceIDIndex{
+							Field: "Service",
+						},
+					},
+				},
+			},
+			"gateway": {
+				Name:         "gateway",
+				AllowMissing: false,
+				Unique:       false,
+				Indexer: &ServiceIDIndex{
+					Field: "Gateway",
+				},
+			},
+			"service": {
+				Name:         "service",
+				AllowMissing: true,
+				Unique:       false,
+				Indexer: &ServiceIDIndex{
+					Field: "Service",
+				},
+			},
+		},
+	}
+}
+
+type ServiceIDIndex struct {
```

Comment:
By the way, I thought this was pretty neat way. Is there a way we can use this Index to deal with `EnterpriseMeta` differences so that we can keep Indexes the same between oss and enterprise?

Reply:
Yeah it will need an enterprise PR that has OSS and Ent implementations for its methods. The method should do some namespace validation.

Will drop a TODO on there so that it doesn't get missed.

Reply:
Great, thanks!